main.cpp

```diff
@@ -102,16 +102,18 @@ int main()
     T t2(20 , "b");                                             //6
     
     Struct1 f;                                            //7
-    auto* smaller = f.compare(&t1 , &t2); 
-    if(smaller != nullptr)                             //8
-        std::cout << "the smaller one is << " << smaller->name << std::endl; //9 
+    auto* smaller = f.compare(t1, t2); 
+    std::cout << "the smaller one is << " << smaller->name << std::endl; //9 
+
+    auto* smaller2 = f.compare2(t1, t2); 
+    std::cout << "the smaller2 one is << " << smaller2->name << std::endl; //9 
```

Comment:
Revert back to the original code here.
The `compare` function is able to return a `nullptr`, which means you need to check if `smaller` is a `nullptr` before you try to use it.
- [ ] revise